### PR TITLE
fix(pnp): esm - use correct error code when module not found

### LINKED ISSUE #5334: [Bug?]: PnP ESM loader throw wrong error code
### Self-service

- [X] I'd be willing to implement a fix

### Describe the bug

If a module is not found while using ESM, the default loader throws error with error code `ERR_MODULE_NOT_FOUND` (as opposed to `MODULE_NOT_FOUND` while using CJS)

However, the PnP ESM loader (`.pnp.loader.mjs`) throws error with code `MODULE_NOT_FOUND`, which breaks compatibility when chaining other loaders. For example, `@esbuild-kit/esm-loader` (and by extension, `tsx`) [checks the error code](https://github.com/esbuild-kit/esm-loader/blob/v2.5.5/src/loaders.ts#L177) to see if it should retry with file extensions.

### To reproduce

```js repro
await packageJsonAndInstall({
    type: 'module'
});

await expect(yarn(
    'node',
    '--input-type=module',
    '-e',
    'import("./a.js").catch((err) => process.stdout.write(err.code))'
)).resolves.toBe('ERR_MODULE_NOT_FOUND')
```

### Environment

```shell
System:
    OS: Windows 10 10.0.19045
    CPU: (16) x64 Intel(R) Core(TM) i7-10700 CPU @ 2.90GHz
  Binaries:
    Node: 19.7.0 - ~\AppData\Local\Temp\xfs-ba57b5d1\node.CMD
    Yarn: 3.5.0 - ~\AppData\Local\Temp\xfs-ba57b5d1\yarn.CMD
    npm: 9.5.0 - C:\Program Files\nodejs\npm.CMD
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by clemyan**
**What's the problem this PR addresses?**

This PR makes the PnP ESM loader throw with error code `ERR_MODULE_NOT_FOUND` (instead of `MODULE_NOT_FOUND`) when a request cannot be resolved.

Closes #5334 

**How did you fix it?**

The error is thrown by the PnP CJS hook, and the ESM loader just calls the CJS hook. So we can just catch the error in the ESM loader, change the error code, and rethrow.

I am not sure about the package releases. I have marked `pnp`, `plugin-pnp` and `cli` as patch releases since other packages doesn't seem to use the ESM loader.

**Checklist**

- [x] I have read the [Contributing Guide](https://yarnpkg.com/advanced/contributing).

- [x] I have set the packages that need to be released for my changes to be effective.

- [x] I will check that all automated PR checks pass before the PR gets reviewed.

### CODE REVIEW

**Review by merceyz — CHANGES_REQUESTED** [review]
Thanks for the fix!

Could you add a test to https://github.com/yarnpkg/berry/blob/ca475757fe85cdc2c4b4441f6d65d38fdab230ad/packages/acceptance-tests/pkg-tests-specs/sources/pnp-esm.test.ts?

**Comment by clemyan** [comment]
Added tests

**Review by merceyz — APPROVED** [review]

File: packages/yarnpkg-pnp/sources/esm-loader/hooks/resolve.ts [hunk 1]
```diff
@@ -100,11 +100,19 @@ export async function resolve(
...
+    result = pnpapi.resolveRequest(specifier, issuer, {
+      conditions: new Set(conditions),
+      // TODO: Handle --experimental-specifier-resolution=node
+      extensions: allowLegacyResolve ? undefined : [],
+    });
+  } catch (err) {
+    if (err instanceof Error && `code` in err && err.code === `MODULE_NOT_FOUND`) {
+      err.code = `ERR_MODULE_NOT_FOUND`;
+      throw err;
+    }
```
Comment (merceyz) [inline packages/yarnpkg-pnp/sources/esm-loader/hooks/resolve.ts:114]:
```suggestion
    if (err instanceof Error && `code` in err && err.code === `MODULE_NOT_FOUND`) {
      err.code = `ERR_MODULE_NOT_FOUND`;
    }
    throw err;
```
We should re-throw the `error` regardless of what it is.

**Review by merceyz — CHANGES_REQUESTED** [review]

File: packages/yarnpkg-pnp/sources/esm-loader/hooks/resolve.ts (hunk 1 above)
Comment (merceyz) [reply packages/yarnpkg-pnp/sources/esm-loader/hooks/resolve.ts]: I pushed the fix and added a test.

**Review by merceyz — APPROVED** [review]
Thanks for the fix!

**Pull request merged**